KAFKA-4144: Allow per stream/table timestamp extractor

jeyhunkarimov opened the pull request.

eliaslevy commented:
👍

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -62,11 +63,14 @@
      * If this is not the case it is the user's responsibility to repartition the date before any key based operation
      * (like aggregation or join) is applied to the returned {@link KStream}.
      *
+     * The default {@link org.apache.kafka.streams.processor.TimestampExtractor} is used as specified in the
```
This should be added to the first paragraph:
```
The default {@code "auto.offset.reset"} strategy, default {@link TimestampExtractor}, and default key and value deserializers as specified in the {@link StreamsConfig config} are used.
```

Please make sure, that the line is not longer than 120 chars.

Please adjust other JavaDocs, too.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -146,7 +166,7 @@
      */
     public <K, V> KStream<K, V> stream(final Serde<K> keySerde, final Serde<V> valSerde, final String... topics) {
-        return stream(null, keySerde, valSerde, topics);
+        return stream((AutoOffsetReset) null, keySerde, valSerde, topics);
```
`return stream(null, null, keySerde, valSerde, topics);`

Do the call directly instead of the cast.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -89,6 +99,7 @@
         return stream(offsetReset, null, null, topics);
```
update to `return stream(offsetReset, null, null, null, topics);` to avoid too many indirections. To this for other overloads, too, please.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -174,7 +197,77 @@
+     * @param topics      the topic names; must contain at least one topic name
+     * @return a {@link KStream} for the specified topics
+     */
+    public <K, V> KStream<K, V> stream(final AutoOffsetReset offsetReset,
```
This should be the only method with actual code. All other overloads should call this one.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -174,7 +197,77 @@
         final String name = newName(KStreamImpl.SOURCE_NAME);
 
-        addSource(offsetReset, name,  keySerde == null ? null : keySerde.deserializer(), valSerde == null ? null : valSerde.deserializer(), topics);
+        addSource(offsetReset, name, null, keySerde == null ? null : keySerde.deserializer(), valSerde == null ? null : valSerde.deserializer(), topics);
```
should not have an implementation but call overloaded method.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -174,7 +197,77 @@
+                                       final Serde<K> keySerde,
+                                       final Serde<V> valSerde,
+                                       final String... topics) {
+        final String name = newName(KStreamImpl.SOURCE_NAME);
```
Should not have an implementation but call overloaded method.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -174,7 +197,77 @@
+     * If this is not the case it is the user's responsibility to repartition the date before any key based operation
+     * (like aggregation or join) is applied to the returned {@link KStream}.
+     *
+     * The source will use specified {@link TimestampExtractor} for extracting timestamps from records.
```
remove this line -- not required.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -174,7 +197,77 @@
+     *
+     * @param offsetReset the {@code "auto.offset.reset"} policy to use for the specified topics if no valid committed
+     *                    offsets are available
+     * @param timestampExtractor the {@link TimestampExtractor} used when consuming records; may be null if the source
```
Nit: adjust indention of other parameters; line should not be longer than 120 chars. Indent second/third/etc line, too.
Text can be shorter:
```
the timestamp extractor used for this source {@link KStream}
if not specified the default extractor defined in the configs will be used
```
(no need to link to `TimestampExtractor` as there will be a link in the JavaDocs anyway.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -201,7 +297,40 @@
+                                       final Serde<K> keySerde,
+                                       final Serde<V> valSerde,
+                                       final Pattern topicPattern) {
+        final String name = newName(KStreamImpl.SOURCE_NAME);
```
as above.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -214,8 +343,48 @@
+     * @param topicPattern the pattern to match for topic names
+     * @return a {@link KStream} for topics matching the regex pattern.
+     */
+    public <K, V> KStream<K, V> stream(final TimestampExtractor timestampExtractor,
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -214,8 +343,48 @@
+     * If this is not the case it is the user's responsibility to repartition the date before any key based operation
+     * (like aggregation or join) is applied to the returned {@link KStream}.
+     *
+     * The source will use specified {@link TimestampExtractor} for extracting timestamps from records.
```
remove

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -214,8 +343,48 @@
+     *
      * @param offsetReset  the {@code "auto.offset.reset"} policy to use for the matched topics if no valid committed
      *                     offsets are available
+     * @param timestampExtractor the {@link TimestampExtractor} used when consuming records; may be null if the source
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -470,6 +837,50 @@
+     * Note that {@link GlobalKTable} always applies {@code "auto.offset.reset"} strategy {@code "earliest"}
+     * regardless of the specified value in {@link StreamsConfig}.
+     *
+     * The source will use specified {@link TimestampExtractor} for extracting timestamps from records.
```
remove

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -470,6 +837,50 @@
+     *
+     * The source will use specified {@link TimestampExtractor} for extracting timestamps from records.
+     *
+     * @param timestampExtractor the {@link TimestampExtractor} used when consuming records; may be null if the source
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -479,16 +890,17 @@
         return new GlobalKTableImpl(new KTableSourceValueGetterSupplier<>(storeName));
     }
 
+
```
No reformatting, please

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -479,16 +890,17 @@
+                valSerde,
+                false,
+                Collections.<String, String>emptyMap(),
+                true).get();
```
No reformatting, please

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java:
```diff
@@ -58,7 +55,7 @@ public RecordQueue queue() {
     private int totalBuffered;
 
-    public PartitionGroup(Map<TopicPartition, RecordQueue> partitionQueues, TimestampExtractor timestampExtractor) {
+    public PartitionGroup(Map<TopicPartition, RecordQueue> partitionQueues) {
```
Nice catch!

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java:
```diff
@@ -185,4 +185,12 @@
      */
     Map<String, Object> appConfigsWithPrefix(String prefix);
 
+    /**
```
We should not add this to the context -- see comments below.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -114,18 +114,18 @@ public StreamTask(TaskId id,
         partitionQueues = new HashMap<>();
 
-        TimestampExtractor timestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
+        //TimestampExtractor timestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
```
keep this but rename to `defaultTimestampExtractor`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -114,18 +114,18 @@ public StreamTask(TaskId id,
         for (TopicPartition partition : partitions) {
             SourceNode source = topology.source(partition.topic());
-            RecordQueue queue = createRecordQueue(partition, source, timestampExtractor);
+            RecordQueue queue = createRecordQueue(partition, source, source.getTimestampExtractor());
```
add line:
`TimestampExtractor sourceTimestampExtractor = source.getTimestampExtractor();`
and change to
`RecordQueue queue = createRecordQueue(partition, source, sourceTimestampExtractor != null ? sourceTimestampExtractor : defaultTimestampExtractor);`

mjsax reviewed (changes requested):
@jeyhunkarimov Thanks for the PR. Looks quite promising.

Put a couple of comments. The comments about `KStreamBuilder#stream()` also apply to `KStreamBuilder#table()` (just did not want to put them twice).

Sorry for being picky with JavaDocs, but we just reworked them and need to keep them clean. (I am not picky about `TopologyBuilder` JavaDocs because those did not get reworked yet :))

mjsax commented:
@jeyhunkarimov One more thing. Please add some tests!

jeyhunkarimov commented:
@mjsax thanks for comments, I will surely take a look.
 I will definitely add tests but before, I wanted to make sure that I am on the right spot. My implementation fails in some tests and I think I know why.

Basically I wanted to get rid of 
`TimestampExtractor timestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG,TimestampExtractor.class);` 
line in `StreamTask` class.
So, I wanted to initialize `timeStampExtractor` property of `SourceNode` in its constructor. If there is per-stream-source `TimeStampExtractor` instance provided, then `SourceNode` will use it, otherwise it will use the one from `StreamConfig`. In this way, we can handle global or per-source based `TimeStampExtractors` inside `SourceNode` class.

So, I modified `StreamConfig` and `ProcessorContext` classes to have `timeStampExtractor` properties. 

The problem is that, when we have application based `TimeStampExtractor` (defined via configuration), then `SourceNode` cannot get it (it gets when accesses `ProcessorContext.timestampExtractor`), and therefore I am getting `NullPointerExceptions`.
 One solution is to pass `StreamConfig` instance to `Topology.build` but I think this is not a nice way to do. 
Another solution would be to keep 
`TimestampExtractor timestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);` line in `StreamTask` class and set `timeStampExtractor` property of `SourceNodes` inside `StreamTask`.

Or any other solution would be welcome.

mjsax commented:
I don't see a good reason why you "wanted to get rid of" this line... Just leave it, as I suggested. Furthermore, semantically the timestamp extractor does not belong to the context, and IMHO it would not be a good design/abstraction adding it their.

jeyhunkarimov commented:
Sorry it is my bad. I realized this short after writing comment and removed comment immediately.

jeyhunkarimov commented:
@mjsax please review

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -53,8 +54,8 @@
      * Create a {@link KStream} from the specified topics.
-     * The default {@code "auto.offset.reset"} strategy and default key and value deserializers as specified in the
-     * {@link StreamsConfig config} are used.
+     * The default {@code "auto.offset.reset"} strategy, default {@link TimestampExtractor} and default key and value
```
Nit: "default {@link TimestampExtractor}[,] and"

The rule is "A and B" (for two things no comma), but "A, B, C, and D" (for three or more things, use commas)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -86,13 +88,14 @@
      * Create a {@link KStream} from the specified topic pattern.
-     * The default {@code "auto.offset.reset"} strategy and default key and value deserializers as specified in the
-     * {@link StreamsConfig config} are used.
+     * The default {@link TimestampExtractor}, default {@code "auto.offset.reset"} strategy and default key and value
```
Nit: Please use the same order in all JavaCode -- above TimestampExtractor is second -- I don't care which order, but please be consistent. Maybe follow parameter order of the method overload that provides all parameters ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -146,12 +151,13 @@
 
     /**
      * Create a {@link KStream} from the specified topics.
+     * The default {@link TimestampExtractor} as specified in the {@link StreamsConfig config} (if any) is used.
```
remove "(if any)"

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -172,9 +178,65 @@
+     * (like aggregation or join) is applied to the returned {@link KStream}.
+     *
+     * @param timestampExtractor the timestamp extractor used for this source {@link KStream},
+     *                           if not specified the default extractor defined in the configs will be used.
```
Nit: no "." at the end
Please update everywhere.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -201,7 +263,7 @@
     public <K, V> KStream<K, V> stream(final Serde<K> keySerde,
```
Update JavaDoc.
same below

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -227,13 +289,70 @@
+     * @param offsetReset        the {@code "auto.offset.reset"} policy to use for the matched topics if no valid
+     *                           committed  offsets are available
+     * @param timestampExtractor the timestamp extractor used for this source {@link KStream},
+     *                           if not specified the default extractor defined in the configs will be used.
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -227,13 +289,70 @@
+     * @param topicPattern       the pattern to match for topic names
+     * @return a {@link KStream} for topics matching the regex pattern.
+     */
+    public <K, V> KStream<K, V> stream(final TimestampExtractor timestampExtractor,
```
Nit: Can you insert this method further down -- we want to order method overloads with regard to number of parameters -- it simplifies to keep track of what overloads are there.

mjsax replied:
Sorry -- mixed it up with `table`.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -114,18 +114,18 @@ public StreamTask(TaskId id,
         for (TopicPartition partition : partitions) {
             SourceNode source = topology.source(partition.topic());
-            RecordQueue queue = createRecordQueue(partition, source, timestampExtractor);
+            RecordQueue queue = createRecordQueue(partition, source, source != null && source.getTimestampExtractor() != null ? source.getTimestampExtractor() : defaultTimestampExtractor);
```
check `source != null` not necessary.
In doubt add an assertion.

jeyhunkarimov replied:
Once I removed it failed most of the tests of `StreamThreadStateStoreProviderTest` class with NullPointerException.

dguy replied:
+1 to what @mjsax said.
The `source` should never be null. So you should change the `StreamThreadStateStoreProviderTest`. It just needs to have the topic name extracted to a field on line 73. And then that same topic name used on line 189 in `new TopicPartition(...)`

dguy replied:
I'd also consider extracting: `source.getTimestampExtractor() != null ? ...` into a local as the line is quite long and it will make the code a bit easier to read.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -608,6 +609,13 @@ public Serde valueSerde() {
     }
 
+
+    public TimestampExtractor timestampExtractor() {
```
This can be reverted.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -178,6 +180,20 @@ public void testProcessOrder() throws Exception {
+        StreamsConfig config = createConfig(baseDir);
+        final TopologyBuilder builder = new TopologyBuilder();
+        builder.addSource(new MockTimestampExtractor(), source1.name(), topic1[0]);
+        builder.addSource(new MockTimestampExtractor2(), source2.name(), topic2[0]);
```
Does this add anything -- I doubt it? (ie, using a second mock TsExtractor)

jeyhunkarimov replied:
Which does not make sense: using two separate tsExtractors or this test case as a whole?

mjsax replied:
The test is fine -- but what's the value in testing overwrite the default extractor two times.

jeyhunkarimov replied:
I see. So I will remove `MockTimestampExtractor2` class and correct the test accordingly.

mjsax reviewed (changes requested):
Thanks for the update.
Most comments are nitty JavaDoc stuff...

@dguy @enothereska Ping for review. Do you think we should add an integration test, too?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -172,9 +178,65 @@
+     * Create a {@link KStream} from the specified topics.
+     * The default {@code "auto.offset.reset"} strategy as specified in the {@link StreamsConfig config} is used.
+     * <p>
+     * If multiple topics are specified there are no ordering guaranteed for records from different topics.
```
guaranteed -> guarantees

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -172,9 +178,65 @@
+    /**
+     * Create a {@link KStream} from the specified topics.
+     * <p>
+     * If multiple topics are specified there are no ordering guaranteed for records from different topics.
```
same as above. I guess this is largely copy & pasted from other javadoc, so the issue is most likely elsewhere

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -301,9 +420,85 @@
+     * The default key and value deserializers as specified in the {@link StreamsConfig config} are used.
+     * Input {@link KeyValue} pairs with {@code null} key will be dropped.
+     * <p>
+     * Note that the specified input topics must be partitioned by key.
```
topics -> topic.

This may well be elsewhere in the java-doc, too

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -172,13 +173,15 @@ public ProcessorNode build() {
+        private final TimestampExtractor timestampExtractor;
 
-        private SourceNodeFactory(String name, String[] topics, Pattern pattern, Deserializer<?> keyDeserializer, Deserializer<?> valDeserializer) {
+        private SourceNodeFactory(String name, String[] topics, Pattern pattern, TimestampExtractor timestampExtractor, Deserializer<?> keyDeserializer, Deserializer<?> valDeserializer) {
```
I know you've only added the one param here, but seeing as you are changing it can you make all the params `final`?

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -347,7 +352,44 @@ public synchronized final TopologyBuilder addSource(String name, String... topic
+     * @param topics             the name of one or more Kafka topics that this source is to consume
+     * @return this builder instance so methods can be chained together; never null
+     */
+    public synchronized final TopologyBuilder addSource(TimestampExtractor timestampExtractor, String name,  String... topics) {
```
Make all params `final`

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -347,7 +352,44 @@ public synchronized final TopologyBuilder addSource(String name, String... topic
+     * @param topics             the name of one or more Kafka topics that this source is to consume
+     * @return this builder instance so methods can be chained together; never null
+     */
+    public synchronized final TopologyBuilder addSource(AutoOffsetReset offsetReset, TimestampExtractor timestampExtractor, String name,  String... topics) {
```
As above

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -381,49 +425,94 @@ public synchronized final TopologyBuilder addSource(String name, Pattern topicPa
+     * @param topicPattern       regular expression pattern to match Kafka topics that this source is to consume
+     * @return this builder instance so methods can be chained together; never null
+     */
+    public synchronized final TopologyBuilder addSource(TimestampExtractor timestampExtractor, String name,  Pattern topicPattern) {
```
As above

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -381,49 +425,94 @@ public synchronized final TopologyBuilder addSource(String name, Pattern topicPa
+     * @param topicPattern       regular expression pattern to match Kafka topics that this source is to consume
+     * @return this builder instance so methods can be chained together; never null
+     */
+    public synchronized final TopologyBuilder addSource(AutoOffsetReset offsetReset, TimestampExtractor timestampExtractor, String name,  Pattern topicPattern) {
```
As above

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -381,49 +425,94 @@ public synchronized final TopologyBuilder addSource(String name, Pattern topicPa
      * @throws TopologyBuilderException if processor is already added or if topics have already been registered by another source
      */
-    public synchronized final TopologyBuilder addSource(AutoOffsetReset offsetReset, String name, Deserializer keyDeserializer, Deserializer valDeserializer, String... topics) {
+    public synchronized final TopologyBuilder addSource(AutoOffsetReset offsetReset,
```
Here also, would be great if you could make the params `final`

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -553,21 +679,27 @@ public synchronized final TopologyBuilder addSource(String name, Deserializer ke
      */
 
-    public synchronized final TopologyBuilder addSource(AutoOffsetReset offsetReset, String name,  Deserializer keyDeserializer, Deserializer valDeserializer, Pattern topicPattern) {
+    public synchronized final TopologyBuilder addSource(AutoOffsetReset offsetReset,
```
And again with `final` if you don't mind

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/SourceNode.java:
```diff
@@ -30,14 +31,21 @@
     private ProcessorContext context;
     private Deserializer<K> keyDeserializer;
     private Deserializer<V> valDeserializer;
+    private TimestampExtractor timestampExtractor;
```
`final` ?
All of the fields should be `final` really

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/SourceNode.java:
```diff
@@ -94,4 +102,7 @@ public String toString(String indent) {
         return sb.toString();
     }
 
+    public TimestampExtractor getTimestampExtractor() {
```
This method can be package-private

mjsax replied:
Can you address this comment, too?

jeyhunkarimov replied:
Because the test classes (`TopologyBuilderTest` for example) cannot access the protected method , I leave it as it is

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -114,18 +114,18 @@ public StreamTask(TaskId id,
         partitionQueues = new HashMap<>();
 
-        TimestampExtractor timestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
+        TimestampExtractor defaultTimestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
```
We should make this `final`, too

dguy commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -178,6 +180,20 @@ public void testProcessOrder() throws Exception {
 
+    @SuppressWarnings("unchecked")
+    @Test
+    public void shouldAddSeparateTimestampExtractorsPerEachSource() throws Exception {
```
maybe `shouldAddTimestampExtractorPerSource` ?

dguy commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -178,6 +180,20 @@ public void testProcessOrder() throws Exception {
+    @SuppressWarnings("unchecked")
+    @Test
+    public void shouldAddSeparateTimestampExtractorsPerEachSource() throws Exception {
+        StreamsConfig config = createConfig(baseDir);
```
Make all locals `final`

dguy reviewed (changes requested):
@jeyhunkarimov - thanks for the PR.

I think there are some tests missing, i.e, for the methods on `KStreamBuilder` and `TopologyBuilder` that add `TimestampExtractor` i don't see any tests for them. We need tests for these in addition to the test added to `StreamTaskTest`

jeyhunkarimov commented:
@dguy 
 - I added extra test for `KStreamBuilder.addSource` with `TimestampExtract` arguments. Not sure it is much different from `TopologyBuilder.addSource` test.
- In `SourceNode` class, not all variables can be final as they are initialized in `init()` method.
- In `StreamTask` 121, the `SourceNode` was getting `null` if the source is defined with `Pattern`.   I deleted "Pattern [ ]" string from `SourceNode`name  (it was initialized like "Pattern [ " + regex + "]" )  so that I can compare with `topic` names.

@mjsax 
I could not get the reason to insert new overloaded methods further below. The next method below is the most generic one (with most arguments) which is 
```
public <K, V> KStream<K, V> stream(final AutoOffsetReset offsetReset,
                                       final TimestampExtractor timestampExtractor,
                                       final Serde<K> keySerde,
                                       final Serde<V> valSerde,
                                       final Pattern topicPattern) 
```

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -114,18 +115,28 @@ public StreamTask(TaskId id,
         for (TopicPartition partition : partitions) {
             SourceNode source = topology.source(partition.topic());
-            RecordQueue queue = createRecordQueue(partition, source, timestampExtractor);
+            // we the source is defined by pattern, we check the sources one by one
```
i'd probably extract lines 121 -> 130 into a method, i.e, `findSourceNode(...)`

Also, we -> if

dguy commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -178,6 +180,26 @@ public void testProcessOrder() throws Exception {
+                restoreStateConsumer, config, new MockStreamsMetrics(new Metrics()), stateDirectory, null, new MockTime(), recordCollector);
+        assertTrue(task.topology().source(topic1[0]).getTimestampExtractor() instanceof MockTimestampExtractor);
+
+        final KStreamBuilder kStreamBuilder = new KStreamBuilder();
```
There is no need to test this as it is calling the same method as above.

dguy commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -178,6 +180,26 @@ public void testProcessOrder() throws Exception {
 
+    @SuppressWarnings("unchecked")
+    @Test
+    public void shouldAddTimestampExtractorPerSource() throws Exception {
```
I'm not sure what this test has to do with `StreamTask`? To me this test should be in `TopologyBuilderTest`. You don't need a `StreamTask` in this case to check that the `TimestampExtractor` was assigned to the source

jeyhunkarimov replied:
I found `StreamTaskTest` the best suitable place, as it was suggested to make `SourceNode.getTimestampExtractor()` method available within package. So, it is not accessible inside `TopologyBuilderTest` currently. 
Then I am making `SourceNode.getTimestampExtractor()`  public and moving the tests to `TopologyBuilderTest`.

mjsax replied:
I agree with @dguy If you test `TopologyBuilder#addSource()` it should go to `TopologyBuilderTest`, and if you test `KStreamBuilder#stream` it should go to `KStreamBuilderTest` -- this also implies, you should split this test into two.

Also testing `KStreamBuilder#addSource` is redundant because its inherited from `TopologyBuilder`.

dguy reviewed (commented):
I'd still like to see some tests for `KStreamBuilder#stream` etc where we are adding the Timestamp extractors.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -86,36 +88,38 @@
+     * deserializers as specified in the {@link StreamsConfig config} are used.
      * <p>
      * If multiple topics are matched by the specified pattern, the created {@link KStream} will read data from all of
      * them and there is no ordering guarantee between records from different topics.
```
I just realized, that we use different wording for topics as array of Strings and topic pattern:
"there is no ordering guarantee" vs "there are no ordering guarantees" -- I think we should clean this up for consistency. Would you mind to add this fix to this PR? The singular version sounds better, IMHO.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -114,18 +115,28 @@ public StreamTask(TaskId id,
-            RecordQueue queue = createRecordQueue(partition, source, timestampExtractor);
+            // we the source is defined by pattern, we check the sources one by one
+            if (source == null) {
+                for (String key: topology.sourceTopics()) {
```
key -> topic

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -114,18 +115,28 @@ public StreamTask(TaskId id,
+        final TimestampExtractor defaultTimestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
 
         for (TopicPartition partition : partitions) {
             SourceNode source = topology.source(partition.topic());
```
I am still confused, about `source` being `null`. In the original code (L121) `source` is handed to `createRecrodQueue` and must not be `null` -- because this was never an issues before, I am still puzzled. why it is now.

jeyhunkarimov replied:
When we add the sources by pattern (`KStreamBuilder.stream(final Pattern topicPattern)`), the source name is given like `"Pattern [" + regex + "]"`. For example, for `t.*` pattern it would be `"Pattern[t.*]"`. In `StreamTask`, we search for sources (in 121) by topic name. For example, for topic name`"topic1"`,  it gives `null`,because  the source name is `"Pattern[t.*]"`.

jeyhunkarimov commented:
@mjsax @dguy would you mind to review again please?

mjsax commented:
Would you mind fixing this typo, too: https://github.com/jeyhunkarimov/kafka/blob/8acf04847fb729155ca72a48e9b29601b781d706/streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java#L829

mjsax commented:
There is a build error. Please fix before we can do a review.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -53,29 +54,30 @@
+     * If multiple topics are specified there is no ordering guarantee for records from different topics.
      * <p>
-     * Note that the specified input topics must be partitioned by key.
+     * Note that the specified input topic must be partitioned by key.
```
revert this for `.stream(...)`, because it can be multiple here. @dguy original comment only applies to `.table()` has has always a single input topic.

jeyhunkarimov commented:
Yes, some PR changed the (test) classes related with this PR so. I will fix them.

mjsax commented:
One more general comment. I would be better if you would not squash your commits on every update -- this can simplify reviewing the PR. And when the PR get's merged, all commits get squashed automatically anyway.

jeyhunkarimov commented:
I fixed the errors

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/KStreamBuilderTest.java:
```diff
@@ -233,4 +237,46 @@ public void shouldMapStateStoresToCorrectSourceTopics() throws Exception {
+    @SuppressWarnings("unchecked")
+    @Test
+    public void shouldAddTimestampExtractorPerSource() throws Exception {
+        TimestampExtractor timestampExtractor = new MockTimestampExtractor();
```
Can you please add `final` wherever possible.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/KStreamBuilderTest.java:
```diff
@@ -233,4 +237,46 @@ public void shouldMapStateStoresToCorrectSourceTopics() throws Exception {
+        TimestampExtractor timestampExtractor = new MockTimestampExtractor();
+        String topicName = "topic";
+        String storeName = "store";
+        KStreamBuilder kStreamBuilder;
```
Why do you not reuse `KStreamBuilderTest#builder` ?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/KStreamBuilderTest.java:
```diff
@@ -233,4 +237,46 @@ public void shouldMapStateStoresToCorrectSourceTopics() throws Exception {
+        kStreamBuilder = new KStreamBuilder();
+        kStreamBuilder.stream(topicName);
+        processorTopology = kStreamBuilder.build(null);
+        assertNull(processorTopology.source(topicName).getTimestampExtractor());
```
It's better to split this test into multiple -- here you test if no source specific extractor is set, thus, this should be a test method `sourceExtractorShouldBeNull` (or similar) and the test should end here.

Apply to below tests, too. (split into positive/negative tests and own tests for stream/table -- for stream/table add overload methods same way to test `TopologBuilder.addSource()`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/TopologyBuilderTest.java:
```diff
@@ -698,4 +699,43 @@ public void shouldSetCorrectSourceNodesWithRegexUpdatedTopics() throws Exception
+    @SuppressWarnings("unchecked")
+    @Test
+    public void shouldAddTimestampExtractorPerSource() throws Exception {
+        TimestampExtractor timestampExtractor = new MockTimestampExtractor();
```
add `final` wherever possible

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -351,7 +351,7 @@ public void process(final Object key, final Object value) {
                                                                  Collections.<StateStore>emptyList());
 
-        final StreamTask streamTask = new StreamTask(new TaskId(0, 0), "applicationId", partitions,
+        final StreamTask streamTask = new StreamTask(new TaskId(0, 0), "applicationId", Utils.mkSet(partition1),
```
why this change?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -145,6 +147,26 @@ public StreamTask(TaskId id,
+        SourceNode source = topology.source(partition.topic());
+        // if the source is defined by pattern, we check the sources one by one
+        if (source == null) {
+            for (String topic: topology.sourceTopics()) {
```
Why not use `topology.sourceTopicPattern()` ? And than check if `partition.topic()` matches the pattern?

jeyhunkarimov replied:
I think `sourceTopicPattern()` is a method of `TopologyBuilder`. In `StreamTask` on the other hand, we get `ProcessorTopology` instance.

mjsax replied:
Ack. By bad.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -172,21 +173,28 @@ public ProcessorNode build() {
             // this should only happen for debugging since during runtime this function should always be called after the metadata has updated.
             if (subscribedTopics.isEmpty())
-                return Collections.singletonList("Pattern[" + pattern + "]");
+                return Collections.singletonList("" + pattern + "");
```
Why this change?

jeyhunkarimov replied:
If we need to find the `topic` of the given source (`StreamTask.findSource()`) by pattern, either we have to remove `"Pattern [ ]"` part from source name and try all matches, or we can remove it (`"Pattern [ ]"` part) when we assign the name for `SourceNode` and directly use its name as `Pattern`. I thought the second case would be more usable.

guozhangwang replied:
@jeyhunkarimov With this change how would the caller differentiate between the case that of a single subscribed topic v.s. a subscribed pattern?

jeyhunkarimov replied:
@guozhangwang actually `"Pattern[ ]"` string is just cosmetic part of the code snippet 
`"Pattern[" + pattern + "]"`. We don't use `"Pattern"` string (anywhere in the code) for detecting if it is a single subscribed topic or subscribed pattern.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -145,6 +147,26 @@ public StreamTask(TaskId id,
+     * @param partition the topic partition
+     * @return
+     */
+    private SourceNode findSource(TopicPartition partition) {
```
apply `final` wherever possible (also within  method)

dguy replied:
+1

mjsax replied:
add `final`

mjsax reviewed (changes requested):
Test code needs some improvements. Overall change looks quite good already.

dguy reviewed (commented):
@mjsax has done a thorough review of this and I concur with his suggestions

jeyhunkarimov commented:
I had to rebase because of conflicts in different commits

mjsax commented:
anyof @dguy @enothereska call for second review

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -116,18 +117,19 @@ public StreamTask(TaskId id,
         for (TopicPartition partition : partitions) {
-            SourceNode source = topology.source(partition.topic());
-            RecordQueue queue = createRecordQueue(partition, source, timestampExtractor);
+            SourceNode source = findSource(partition);
```
It would be nice if you made these `final` while you are doing this change.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -147,6 +149,26 @@ public StreamTask(TaskId id,
+        if (source == null) {
+            for (String topic: topology.sourceTopics()) {
+                if (Pattern.compile(topic).matcher(partition.topic()).matches()) {
+                    source = topology.source(topic);
```
`return topology.source(topic);`

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -147,6 +149,26 @@ public StreamTask(TaskId id,
+     * @return
+     */
+    private SourceNode findSource(TopicPartition partition) {
+        SourceNode source = topology.source(partition.topic());
```
`final`

dguy commented on streams/src/test/java/org/apache/kafka/streams/kstream/KStreamBuilderTest.java:
```diff
@@ -233,4 +236,54 @@ public void shouldMapStateStoresToCorrectSourceTopics() throws Exception {
+        builder.stream(new MockTimestampExtractor(), null, null, "topic");
+        final ProcessorTopology processorTopology = builder.build(null);
+        for (final SourceNode sourceNode: processorTopology.sources()) {
+            assertTrue(sourceNode.getTimestampExtractor() instanceof MockTimestampExtractor);
```
using `assertThat` is nicer as it gives better failure messages.
`assertThat(sourceNode.getTimestampExtractor(), instanceOf(MockTimestaampExtractor))`

in other places, too

dguy commented on streams/src/test/java/org/apache/kafka/streams/kstream/KStreamBuilderTest.java:
```diff
@@ -233,4 +236,54 @@ public void shouldMapStateStoresToCorrectSourceTopics() throws Exception {
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void kStreamhouldAddTimestampExtractorWithOffsetResetPerSource() throws Exception {
```
typo: kStreamhould... -> kStreamShould
In fact i'd probably rename these methods to begin with should, i.e., 
`shouldAddTimestampExtractorToStreamWithOffsetResetPerSource` etc

dguy commented on streams/src/test/java/org/apache/kafka/streams/processor/TopologyBuilderTest.java:
```diff
@@ -698,4 +699,61 @@ public void shouldSetCorrectSourceNodesWithRegexUpdatedTopics() throws Exception
+        final TopologyBuilder builder = new TopologyBuilder();
+        builder.addSource(new MockTimestampExtractor(), "source", "topic");
+        final ProcessorTopology processorTopology = builder.build(null);
+        assertTrue(processorTopology.source("topic").getTimestampExtractor() instanceof MockTimestampExtractor);
```
as per previous `assertThat(..., instanceOf(...))` would be better

jeyhunkarimov commented:
I did changes

dguy reviewed (approved):
Thanks @jeyhunkarimov LGTM

mjsax commented:
I just realized, that we do have a KIP for this. I am very sorry that I missed to mention this from the beginning on.

@jeyhunkarimov Are you familiar with the KIP process? Would you like to do the KIP?

See https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals

jeyhunkarimov commented:
@mjsax np. In general, I am familiar with KIP process
So is there some existing KIP related with this PR (if yes, can you provide the number) or should I create KIP for this PR and discuss in mailing list?

mjsax commented:
We don't have a KIP. Please, use the template to create a new KIP page, add it to the table "KIPs under discussion" (and to Kafka Streams page in the Wiki) and start a DISCUSS thread on the dev mailing list. Thanks a lot!

jeyhunkarimov commented:
I don't have `Create` button on my confluence page (https://cwiki.apache.org/confluence/display/~jeyhun.karimov)
Do I have to ask somebody to give me authorization to create KIP?

mjsax commented:
@guozhangwang Can you give write access to @jeyhunkarimov to the Kafka wiki so he can prepare the KIP ?

guozhangwang commented:
@jeyhunkarimov what is your apache id?

mjsax commented:
jeyhun.karimov
see link in Jeyhun's comment :)

guozhangwang commented:
Oh right. Done.

mjsax commented:
This PR need a further update (cf. the JIRA -- I just updated it). Also see my mail on the dev list.

mjsax commented:
Ignore my last comment (deleted it already)... I did update the JIRA, but it does not require a change of this PR.

@jeyhunkarimov You might want to sent another reminder to dev list about the KIP. Ask for further feedback and if nodoby response, you might want to start the VOTE thread.

mjsax commented:
Retest this please.

mjsax commented:
The test errors are a little weird. The test that fails is not part of you current PR branch but was added to `trunk` later on. Can you please rebase to `trunk` -- I hope this fixed the issue. Not sure why it occurred in the first place.

jeyhunkarimov commented:
The tests will fail when there are no source nodes defined in the test topology. Also, if the defined source nodes are not related with partitions, the tests will also fail due to the NullPointerException in `StreamTask` 125: 
               
   ```
final SourceNode source = findSource(partition);   // <- here.  if partitions and source nodes don't match source will be null
final TimestampExtractor sourceTimestampExtractor = source.getTimestampExtractor() != null ? source.getTimestampExtractor() : defaultTimestampExtractor;
```


Maybe we should add extra check specifically for that issue, when validating the topology.

mjsax commented:
> The tests will fail when there are no source nodes defined in the test topology.

I think that's fine -- for tests it would be a testing error to not add a source. Not sure if we have a proper test for real topologies to guard against it (only required for PAPI) -- for KStreamBuilder it cannot happen. Maybe `TopologyBuilder` should contain a test that ensures that the DAG is connected and "well-formed" ? But I actually think, we would not even generate the `ProcessorTopology` in the first place -- we start building it from the sources, and if no source was added, the whole `ProcessorTopology` would be empty.

> Also, if the defined source nodes are not related with partitions, the tests will also fail due to the NullPointerException in StreamTask 125:

This would also be a testing error only. For real topologies, if there are not partitions, we would not create a task for the `SourceNode` in the first place and thus, this code should never be executed.

\cc @enothereska @dguy @guozhangwang (does this make sense?)

jeyhunkarimov commented:
Sorry I had to rebase again, because of conflicts.

mjsax commented:
No worries @jeyhunkarimov For rebasing, you actually don't need to squash previous commit (but I know that rebasing is simple if you have a single commit). However, you could squash and rebase before you do the PR update and put a new commit with the change on top :)

mjsax reviewed (approved):
LGTM. @guozhangwang please review and merge after VOTE on KIP-123 passed.

jeyhunkarimov commented:
Sorry, I had to rebase again as there were a lot of conflicts among my commits and among mine and other commits. I deprecated  
```
StreamConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG
StreamConfig.TIMESTAMP_EXTRACTOR_CLASS_DOC
StreamConfig.KEY_SERDE_CLASS_CONFIG
StreamConfig.KEY_SERDE_CLASS_DOC
StreamConfig.VALUE_SERDE_CLASS_CONFIG
StreamConfig.VALUE_SERDE_CLASS_DOC
StreamConfig.keySerde()
StreamConfig.valueSerde()
```

For test classes, I replaced the deprecated variables with new ones. For non-test classes, I put an if condition to check both of the variables.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -131,8 +131,14 @@
     public static final String TIMESTAMP_EXTRACTOR_CLASS_CONFIG = "timestamp.extractor";
-    private static final String TIMESTAMP_EXTRACTOR_CLASS_DOC = "Timestamp extractor class that implements the <code>TimestampExtractor</code> interface.";
+    @Deprecated
+    private static final String TIMESTAMP_EXTRACTOR_CLASS_DOC = "Timestamp extractor class that implements the <code>TimestampExtractor</code> interface. This config is deprecated, use {@link } instead";
```
please fix this: `use {@link } instead`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -555,10 +589,11 @@ public StreamsConfig(final Map<?, ?> props) {
     /**
      * Return an {@link Serde#configure(Map, boolean) configured} instance of {@link #KEY_SERDE_CLASS_CONFIG key Serde
-     * class}.
+     * class}. This method is deprecated. Use {@link #defaultKeySerde()} method instead
```
Nit: add missing `.` at the end.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -570,11 +605,28 @@ public Serde keySerde() {
+    /**
      * Return an {@link Serde#configure(Map, boolean) configured} instance of {@link #VALUE_SERDE_CLASS_CONFIG value
-     * Serde class}.
+     * Serde class}. This method is deprecated. Use {@link #defaultValueSerde()} instead
```
Nit: missing `.`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -115,17 +116,21 @@ public StreamTask(TaskId id,
-        TimestampExtractor timestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
-
+        TimestampExtractor deprecatedTimestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
+        TimestampExtractor nonDeprecatedTimestampExtractor = config.getConfiguredInstance(StreamsConfig.DEFAULT_TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
```
Nit: add `final` twice

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -115,17 +116,21 @@ public StreamTask(TaskId id,
+        final TimestampExtractor defaultTimestampExtractor  = nonDeprecatedTimestampExtractor == null ? deprecatedTimestampExtractor : nonDeprecatedTimestampExtractor;
+        
+        // if timestampextrator is null, check also with non-deprecated
         for (TopicPartition partition : partitions) {
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -115,17 +116,21 @@ public StreamTask(TaskId id,
-            RecordQueue queue = createRecordQueue(partition, source, timestampExtractor);
+            final SourceNode source = findSource(partition);
+            final TimestampExtractor sourceTimestampExtractor = source.getTimestampExtractor() != null ? source.getTimestampExtractor() : defaultTimestampExtractor;
+            RecordQueue queue = createRecordQueue(partition, source, sourceTimestampExtractor);
```
Nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -145,6 +150,25 @@ public StreamTask(TaskId id,
+        final SourceNode source = topology.source(partition.topic());
+        // if the source is defined by pattern, we check the sources one by one
+        if (source == null) {
+            for (String topic: topology.sourceTopics()) {
```
Nit: add `final`

mjsax reviewed (changes requested):
@jeyhunkarimov Thanks for rebasing.

Can you also update JavaDocs for `TimestampExtractor` interface to explain that an implementation must be stateless. Thx.

jeyhunkarimov commented:
@mjsax can you please check?

mjsax reviewed (approved):
LGTM

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -257,24 +275,39 @@
             .define(TIMESTAMP_EXTRACTOR_CLASS_CONFIG,
                     Type.CLASS,
-                    FailOnInvalidTimestamp.class.getName(),
+                    null,
```
For backward compatibility, we need to keep the old default value.
Btw: we don't do any ordering here yet -- just above. so no need to reorder anything here.

mjsax replied:
For backward compatibility, we need to keep the old default value.
Btw: we don't do any ordering here yet -- just above. so no need to reorder anything here.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -125,8 +139,27 @@
+    private static final String DEFAULT_VALUE_SERDE_CLASS_DOC = "Default serializer / deserializer class for value that implements the <code>Serde</code> interface.";
+
+    /** {@code metrics.sample.window.ms} */
+    public static final String METRICS_SAMPLE_WINDOW_MS_CONFIG = CommonClientConfigs.METRICS_SAMPLE_WINDOW_MS_CONFIG;
```
Just some nitpick: we started to to order all configs alphabetically here -- it make it a little simpler to keep an overview and to maintain the code. Would you mind to not move configs that get deprecate and add the new config at the "right" place. Thanks a lot. :)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -257,24 +275,39 @@
             .define(VALUE_SERDE_CLASS_CONFIG,
                     Type.CLASS,
-                    Serdes.ByteArraySerde.class.getName(),
+                    null,
```
As above: need to keep default value.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -580,37 +613,77 @@ public StreamsConfig(final Map<?, ?> props) {
-            serde.configure(originals(), true);
+            Serde<?> serde = getConfiguredInstance(KEY_SERDE_CLASS_CONFIG, Serde.class);
+            if (serde == null) {
+                serde = getConfiguredInstance(DEFAULT_KEY_SERDE_CLASS_CONFIG, Serde.class);
```
you can simple call `serde = defaultKeySerde()` here.

mjsax replied:
you can simple call `serde = defaultKeySerde()` here.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -580,37 +613,77 @@ public StreamsConfig(final Map<?, ?> props) {
+    public Serde defaultKeySerde() {
+        try {
+            final Serde<?> serde = getConfiguredInstance(DEFAULT_KEY_SERDE_CLASS_CONFIG, Serde.class);
+            serde.configure(originals(), true);
```
`.configure()` is called within `getConfiguredInstance()` already -- you can remove this line (I know this pattern was there before, but it's wrong -- can you please fit it :))

this can be a single liner within try-catch-block: `return getConfiguredInstance(DEFAULT_KEY_SERDE_CLASS_CONFIG, Serde.class);`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -580,37 +613,77 @@ public StreamsConfig(final Map<?, ?> props) {
-            serde.configure(originals(), false);
+            Serde<?> serde = getConfiguredInstance(VALUE_SERDE_CLASS_CONFIG, Serde.class);
+            if (serde == null) {
+                serde = getConfiguredInstance(DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serde.class);
```
As above

mjsax replied:
As above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -580,37 +613,77 @@ public StreamsConfig(final Map<?, ?> props) {
+     */
+    public Serde defaultValueSerde() {
+        try {
+            final Serde<?> serde = getConfiguredInstance(DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serde.class);
```
As above.

mjsax replied:
As above.

mjsax reviewed (commented):
I did one more pass, and do have some further minor comments.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -125,8 +135,24 @@
+
+    /** {@code default value.serde} */
+    public static final String DEFAULT_VALUE_SERDE_CLASS_CONFIG = "default.value.serde";
+    private static final String DEFAULT_VALUE_SERDE_CLASS_DOC = "Default serializer / deserializer class for value that implements the <code>Serde</code> interface.";
```
Just some nitpick: we started to to order all configs alphabetically here -- it make it a little simpler to keep an overview and to maintain the code. Would you mind to not move configs that get deprecate and add the new config at the "right" place. Thanks a lot. :)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -257,24 +275,39 @@
             .define(VALUE_SERDE_CLASS_CONFIG,
                     Type.CLASS,
-                    Serdes.ByteArraySerde.class.getName(),
+                    null,
```
As above: need to keep default value.

jeyhunkarimov replied:
If we want to distinguish between `VALUE_SERDE_CLASS_CONFIG` and `DEFAULT_VALUE_SERDE_CLASS_CONFIG` for example, in     `public Serde valueSerde()`  method, we have to differentiate whether 

` Serde<?> serde = getConfiguredInstance(VALUE_SERDE_CLASS_CONFIG, Serde.class);
`
is default or not. If it is default, then we will call `defaultValueSerde()` method. If we don't set the default value to `null`, then its default value will be some Object initialized with `Serdes.ByteArraySerde.class.getName()` class. In this case, it is hard to know whether the old `serde` (`VALUE_SERDE_CLASS_CONFIG`) is overridden (so we return it) or it has  default value (so we call `defaultValueSerde() `method). 

Moreover, because we are controlling the access to `VALUE_SERDE_CLASS_CONFIG` via `valueSerde()` method, we handle `null` cases here as well.

The same applies to `KEY_SERDE_CLASS_CONFIG` and `TIMESTAMP_EXTRACTOR_CLASS_CONFIG` as well.

mjsax replied:
Ah. Makes sense!

mjsax replied:
It's actually pretty elegant! :)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -580,37 +613,77 @@ public StreamsConfig(final Map<?, ?> props) {
+    public Serde defaultKeySerde() {
+        try {
+            final Serde<?> serde = getConfiguredInstance(DEFAULT_KEY_SERDE_CLASS_CONFIG, Serde.class);
+            serde.configure(originals(), true);
```
`.configure()` is called within `getConfiguredInstance()` already -- you can remove this line (I know this pattern was there before, but it's wrong -- can you please fit it :))

this can be a single liner within try-catch-block: `return getConfiguredInstance(DEFAULT_KEY_SERDE_CLASS_CONFIG, Serde.class);`

jeyhunkarimov replied:
I think in `getConfiguredInstance() ` method, `Configurable.configure(Map<String, ?> configs)` method is called, but in `keySerde() `and `valueSerde() `methods we need also to call `Serde.configure(Map<String, ?> configs, boolean isKey)` method.  So I think the two `configure` methods are different.

mjsax replied:
`Serde` implements `Configurable` -- the object `o` in `((Configurable) o).configure(originals());` is the serde object.

jeyhunkarimov replied:
Yes I want to say that `Serde.configure(arg1, arg2)` is different from `Configurable.configure(arg1)`. So, inside `getConfiguredInstance()` method it is not called `configure(arg1, arg2)` but  `configure(arg1)`

mjsax replied:
I did have a closer look into the code. You are right. I also double checked and `Serde` does actually not implement `Configurable` (so there will also not be two calls what would be bad).

Sorry for the confusion -- and thanks a lot for pointing out that it is correct as is!!

mjsax reviewed (changes requested):
Thanks for rebasing! I did one more pass and have some more minor comments. Overall very nice work!

mjsax commented:
@jeyhunkarimov Can you rebase once again -- we merged a few conflicting PR today. Sorry for that.

Call for review @enothereska @dguy

enothereska commented:
@guozhangwang could we get this in please? Thanks.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -115,15 +116,18 @@ public StreamTask(final TaskId id,
 
-        final TimestampExtractor timestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
-
+        final TimestampExtractor deprecatedTimestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
```
Could we add a similar function in `StreamsConfig` as `keySerde` in which we capture the deprecated / new default values there so that we do not need to leak that logic here?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/SourceNode.java:
```diff
@@ -29,14 +30,20 @@
+    private final TimestampExtractor timestampExtractor;
 
-    public SourceNode(String name, List<String> topics, Deserializer<K> keyDeserializer, Deserializer<V> valDeserializer) {
+    public SourceNode(String name, List<String> topics, TimestampExtractor timestampExtractor, Deserializer<K> keyDeserializer, Deserializer<V> valDeserializer) {
```
Related to the question I have before: in the `toString` function at line 85, we are printing `topics:string`, and we canno tell if it is a pattern or a single topic right?

jeyhunkarimov replied:
Yes, once the source is defined and added to topology, we cannot tell if it is defined by name or by pattern. This was the case before this PR as well.

guozhangwang replied:
Thanks for the explanation, makes sense.

jeyhunkarimov replied:
I recognized this issue:

        `builder.stream(Pattern.compile("t.*"))`;  
// this line adds sourceNode with name "t*", (before this PR it was `"Pattern[t*]"`)
       After that, in `StreamTask:120` when one executes:

         `final SourceNode source = topology.source(partition.topic());`  
 // where `partition.topic()` is "t1"

Here the `source` will be  `null` because  `topology.source()` is just key value lookup. 
So  I moved extra extra lookup for patterns from `StreamTask` to` topology.source()`

guozhangwang replied:
@jeyhunkarimov Hmm, in `TopologyBuilder#build` when we are adding the source node we will execute the pattern matching from the current topic metadata so the map should already be filled with the actual topic right?

I think the problem, as I added in `SourceNodeFactory#getTopics`'s comment, is that under debugging / unit testing environment, there is no topic metadata available to pass the topic list to `List<String> getTopics(Collection<String> subscribedTopics)`. What we need to do to fix the unit tests then is to call `SubscriptionUpdates#updateTopics()` before calling `builder.build()` for regex involved tests.

```
                    final List<String> topics = (sourceNodeFactory.pattern != null) ?
                            sourceNodeFactory.getTopics(subscriptionUpdates.getUpdates()) :
                            sourceNodeFactory.topics;

                    for (String topic : topics) {
                        if (internalTopicNames.contains(topic)) {
                            // prefix the internal topic name with the application id
                            topicSourceMap.put(decorateTopic(topic), (SourceNode) node);
                        } else {
                            topicSourceMap.put(topic, (SourceNode) node);
                        }
                    }
```

bbejeck replied:
@jeyhunkarimov Sorry for jumping in a little late.  @guozhangwang  is correct about the test.

An example for unit-testing with regex defined topics using the `TopologyBuilder` is `TopologyBuilderTest#shouldSetCorrectSourceNodesWithRegexUpdatedTopics` (line 675)

jeyhunkarimov replied:
@guozhangwang @bbejeck thanks for your comments.
Done.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -142,6 +146,25 @@ public StreamTask(final TaskId id,
+     * @param partition the topic partition
+     * @return
+     */
+    private SourceNode findSource(final TopicPartition partition) {
```
Why do we need to augment this function here? I.e. by the time the stream task is created, we should have populated the `sourceByTopics` map with the pattern matched topics already, so I'm not sure if the additional computational logic is needed? cc @bbejeck .

jeyhunkarimov replied:
@guozhangwang You are right. The main reason that I put this extra function is that I was getting many fails in tests. 

There were some tests with no source nodes defined in the test topology  or the defined source nodes are not related with partitions.

guozhangwang replied:
Hmm, maybe it's better to fix these tests than modifying the production-code? E.g. we can add some test-only function to fill the topics from the pattern in `@before` rule.

jeyhunkarimov replied:
@guozhangwang I fixed the tests as well and added this code snippet too. 
I mentioned this in 
[here](https://github.com/apache/kafka/pull/2466#issuecomment-282328998)

jeyhunkarimov replied:
But I got your point

guozhangwang replied:
Yeah I think if the logic is not needed in real cases then we should not add that since it may hide some potential bugs. As long as the tests can be covered then we can get rid of the unnecessary logic.

guozhangwang reviewed (commented):
Thanks for the patch @jeyhunkarimov . LGTM overall, just a couple of comments. Also could you please rebase so we can merge it before the feature freeze deadline?

jeyhunkarimov committed:
all conflicts resolved while rebasing

another set of conflicts resolved

all conflicts resolved while rebasing

all conflicts resolved while rebasing

unit tests fixed

guozhangwang reviewed (approved):
Merged to trunk. Thanks @jeyhunkarimov for the big contribution!
